[Bug]: db connection leak when sql exception ?

### Apache Hop version?

2.15.0

### Java version?

17

### Operating system

Windows

### What happened?

I loop run Table input to query multi sql,  db is oracle
and I found when sql is  exception the connection will be not release (close?) 
and I face **IO Error: Got minus one from a read call**  (reach max connection?)
and that hop jvm connot connection to db forever

the problem maybe is in org.apache.hop.core.database.Database.openQuery() no close connection when  sql exception ?

### Issue Priority

Priority: 2

### Issue Component

Component: Database

the error is Caused by:

```
Unable to get queryfields for SQL: 
SELECT <values> FROM <table name> WHERE <conditions>

Couldn't get field info from [SELECT <values> FROM <table name> WHERE <conditions>]

ORA-00936: missing expression

https://docs.oracle.com/error-help/db/ora-00936/


	at org.apache.hop.pipeline.transforms.tableinput.TableInputMeta.getFields(TableInputMeta.java:241)
	at org.apache.hop.pipeline.transform.BaseTransformMeta.getFields(BaseTransformMeta.java:328)
	at org.apache.hop.pipeline.PipelineMeta.getThisTransformFields(PipelineMeta.java:1414)
	at org.apache.hop.pipeline.PipelineMeta.getTransformFields(PipelineMeta.java:1222)
	at org.apache.hop.pipeline.PipelineMeta.getTransformFields(PipelineMeta.java:1116)
	at org.apache.hop.pipeline.PipelineMeta.getTransformFields(PipelineMeta.java:1082)
	at org.apache.hop.pipeline.engines.local.LocalPipelineEngine.addTransformExecutionSamplers(LocalPipelineEngine.java:323)
	at org.apache.hop.pipeline.engines.local.LocalPipelineEngine.prepareExecution(LocalPipelineEngine.java:242)
	at org.apache.hop.pipeline.Pipeline.execute(Pipeline.java:528)
	at org.apache.hop.workflow.actions.pipeline.ActionPipeline.execute(ActionPipeline.java:543)
	at org.apache.hop.workflow.Workflow.executeFromStart(Workflow.java:750)
	at org.apache.hop.workflow.Workflow.executeFromStart(Workflow.java:900)
	at org.apache.hop.workflow.Workflow.executeFromStart(Workflow.java:451)
	at org.apache.hop.workflow.Workflow.startExecution(Workflow.java:311)
	at org.apache.hop.workflow.engines.local.LocalWorkflowEngine.startExecution(LocalWorkflowEngine.java:249)
	at org.apache.hop.pipeline.transforms.workflowexecutor.WorkflowExecutor.executeWorkflow(WorkflowExecutor.java:223)
	at org.apache.hop.pipeline.transforms.workflowexecutor.WorkflowExecutor.processRow(WorkflowExecutor.java:177)
	at org.apache.hop.pipeline.transform.RunThread.run(RunThread.java:54)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.apache.hop.core.exception.HopDatabaseException: 
Couldn't get field info from [SELECT <values> FROM <table name> WHERE <conditions>]

ORA-00936: missing expression

https://docs.oracle.com/error-help/db/ora-00936/

	at org.apache.hop.core.database.Database.getQueryFieldsFallback(Database.java:2327)
	at org.apache.hop.core.database.Database.getQueryFields(Database.java:2153)
	at org.apache.hop.pipeline.transforms.tableinput.TableInputMeta.getFields(TableInputMeta.java:230)
	... 18 more
Caused by: java.sql.SQLSyntaxErrorException: ORA-00936: missing expression

https://docs.oracle.com/error-help/db/ora-00936/
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:715)
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:615)
	at oracle.jdbc.driver.T4C8Oall.processError(T4C8Oall.java:1315)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:969)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:237)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:508)
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:285)
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:1145)
	at oracle.jdbc.driver.OracleStatement.prepareDefineBufferAndExecute(OracleStatement.java:1300)
	at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe(OracleStatement.java:1225)
	at oracle.jdbc.driver.OracleStatement.executeSQLSelect(OracleStatement.java:1593)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1443)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3950)
	at oracle.jdbc.driver.OraclePreparedStatement.executeQuery(OraclePreparedStatement.java:4054)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery(OraclePreparedStatementWrapper.java:1113)
	at oracle.ucp.jdbc.proxy.oracle$1ucp$1jdbc$1proxy$1oracle$1StatementProxy$2oracle$1jdbc$1internal$1OraclePreparedStatement$$$Proxy.executeQuery(Unknown Source)
	at org.apache.hop.core.database.Database.getQueryFieldsFallback(Database.java:2304)
	... 20 more
Caused by: Error : 936, Position : 7, SQL = SELECT <values> FROM <table name> WHERE <conditions>, Original SQL = SELECT <values> FROM <table name> WHERE <conditions>, Error Message = ORA-00936: missing expression

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:723)
	... 36 more
```

--- a/engine/src/main/java/org/apache/hop/pipeline/Pipeline.java
+++ b/engine/src/main/java/org/apache/hop/pipeline/Pipeline.java
@@ -1151,19 +1151,7 @@ public abstract class Pipeline
       // Also explicitly call dispose() to clean up resources opened during
       // init()
       //
-      for (TransformInitThread initThread : initThreads) {
-        TransformMetaDataCombi combi = initThread.getCombi();
-
-        // Dispose will overwrite the status, but we set it back right after
-        // this.
-        combi.transform.dispose();
-
-        if (initThread.isOk()) {
-          combi.data.setStatus(ComponentExecutionStatus.STATUS_HALTED);
-        } else {
-          combi.data.setStatus(ComponentExecutionStatus.STATUS_STOPPED);
-        }
-      }
+      disposeInitializedTransforms();
 
       // Just for safety, fire the pipeline finished listeners...
       try {
@@ -1196,6 +1184,33 @@ public abstract class Pipeline
     log.snap(Metrics.METRIC_PIPELINE_INIT_STOP);
 
     setReadyToStart(true);
+  }
+
+  /**
+   * Dispose the transforms which were initialized in {@link #prepareExecution()}. This releases the
+   * resources they acquired during init() (database connections, files, ...) in the situations
+   * where the transform threads are never started and {@link RunThread} can't do it for us.
+   */
+  public void disposeInitializedTransforms() {
+    if (transforms == null) {
+      return;
+    }
+    for (TransformMetaDataCombi combi : transforms) {
+      // Dispose will overwrite the status, but we set it back right after this.
+      //
+      ComponentExecutionStatus status = combi.data.getStatus();
+      try {
+        combi.transform.dispose();
+      } catch (Exception e) {
+        // A transform which fails to clean up shouldn't keep the others from doing so.
+        //
+        log.logError("Error disposing transform " + combi.transformName, e);
+      }
+      combi.data.setStatus(
+          status == ComponentExecutionStatus.STATUS_STOPPED
+              ? ComponentExecutionStatus.STATUS_STOPPED
+              : ComponentExecutionStatus.STATUS_HALTED);
+    }
   }
 
   /**

--- a/engine/src/main/java/org/apache/hop/pipeline/engines/local/LocalPipelineEngine.java
+++ b/engine/src/main/java/org/apache/hop/pipeline/engines/local/LocalPipelineEngine.java
@@ -235,17 +235,37 @@ public class LocalPipelineEngine extends Pipeline implements IPipelineEngine<Pip
 
     super.prepareExecution();
 
-    // Do the lookup of the execution information only once
-    lookupExecutionInformationLocation();
-
-    // Register the pipeline
-    registerPipelineExecutionInformation();
-
-    // Attach samplers to the transforms if needed
+    // The transforms are initialized at this point, which means they can hold on to database
+    // connections, files, ... If anything below fails we never get to start the transform threads,
+    // so nothing would ever dispose them. Clean up explicitly to avoid leaking those resources.
     //
-    addTransformExecutionSamplers();
+    try {
+      // Do the lookup of the execution information only once
+      lookupExecutionInformationLocation();
 
-    //
+      // Register the pipeline
+      registerPipelineExecutionInformation();
+
+      // Attach samplers to the transforms if needed
+      //
+      addTransformExecutionSamplers();
+    } catch (Exception e) {
+      disposeInitializedTransforms();
+
+      // Just for safety, fire the pipeline finished listeners...
+      //
+      try {
+        fireExecutionFinishedListeners();
+      } catch (HopException listenerException) {
+        log.logError(
+            "Error firing the execution finished listeners after a failed preparation",
+            listenerException);
+      } finally {
+        // Flag the pipeline as finished even if an exception was thrown
+        setFinished(true);
+      }
+      throw e;
+    }
   }
 
   /**

--- a/engine/src/test/java/org/apache/hop/pipeline/engines/local/LocalPipelineEngineTest.java
+++ b/engine/src/test/java/org/apache/hop/pipeline/engines/local/LocalPipelineEngineTest.java
@@ -1,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hop.pipeline.engines.local;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+import org.apache.hop.core.HopEnvironment;
+import org.apache.hop.core.exception.HopException;
+import org.apache.hop.core.logging.ILogChannel;
+import org.apache.hop.junit.rules.RestoreHopEngineEnvironmentExtension;
+import org.apache.hop.pipeline.PipelineMeta;
+import org.apache.hop.pipeline.engine.EngineComponent.ComponentExecutionStatus;
+import org.apache.hop.pipeline.transform.BaseTransformData;
+import org.apache.hop.pipeline.transform.ITransform;
+import org.apache.hop.pipeline.transform.ITransformData;
+import org.apache.hop.pipeline.transform.TransformMetaDataCombi;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(RestoreHopEngineEnvironmentExtension.class)
+class LocalPipelineEngineTest {
+
+  @BeforeAll
+  static void beforeClass() throws HopException {
+    HopEnvironment.init();
+  }
+
+  /**
+   * Transforms acquire resources (database connections, files, ...) in init(). When the preparation
+   * of the pipeline fails after that init() the transform threads are never started, so nothing
+   * else will ever release them. See issue #5970.
+   */
+  @Test
+  void transformsAreDisposedWhenPreparationFailsAfterInit() {
+    LocalPipelineEngine pipeline =
+        spy(
+            new LocalPipelineEngine(new PipelineMeta()) {
+              @Override
+              public void addTransformExecutionSamplers() throws HopException {
+                throw new HopException("Unable to attach the samplers");
+              }
+            });
+    pipeline.setLogChannel(mock(ILogChannel.class));
+
+    HopException exception = assertThrows(HopException.class, pipeline::prepareExecution);
+
+    assertTrue(exception.getMessage().contains("Unable to attach the samplers"));
+    verify(pipeline).disposeInitializedTransforms();
+    assertTrue(pipeline.isFinished(), "The pipeline should be flagged as finished");
+  }
+
+  @Test
+  void disposeInitializedTransformsDisposesEveryTransform() throws Exception {
+    LocalPipelineEngine pipeline = new LocalPipelineEngine(new PipelineMeta());
+    pipeline.setLogChannel(mock(ILogChannel.class));
+    pipeline.prepareExecution();
+
+    // One transform which fails to clean up shouldn't keep the others from doing so.
+    //
+    TransformMetaDataCombi failing =
+        combi("failing", ComponentExecutionStatus.STATUS_IDLE, new RuntimeException("Oops"));
+    TransformMetaDataCombi initialized =
+        combi("initialized", ComponentExecutionStatus.STATUS_IDLE, null);
+    TransformMetaDataCombi notInitialized =
+        combi("not-initialized", ComponentExecutionStatus.STATUS_STOPPED, null);
+    pipeline.getTransforms().add(failing);
+    pipeline.getTransforms().add(initialized);
+    pipeline.getTransforms().add(notInitialized);
+
+    pipeline.disposeInitializedTransforms();
+
+    verify(failing.transform).dispose();
+    verify(initialized.transform).dispose();
+    verify(notInitialized.transform).dispose();
+
+    assertEquals(ComponentExecutionStatus.STATUS_HALTED, failing.data.getStatus());
+    assertEquals(ComponentExecutionStatus.STATUS_HALTED, initialized.data.getStatus());
+    assertEquals(
+        ComponentExecutionStatus.STATUS_STOPPED,
+        notInitialized.data.getStatus(),
+        "A transform which never initialized should keep its stopped status");
+  }
+
+  private TransformMetaDataCombi combi(
+      String name, ComponentExecutionStatus status, RuntimeException disposeError) {
+    ITransform transform = mock(ITransform.class);
+    if (disposeError != null) {
+      doThrow(disposeError).when(transform).dispose();
+    }
+    ITransformData data = new BaseTransformData() {};
+    data.setStatus(status);
+
+    TransformMetaDataCombi combi = new TransformMetaDataCombi();
+    combi.transformName = name;
+    combi.transform = transform;
+    combi.data = data;
+    return combi;
+  }
+}

